### DISCUSSION TITLE
Use a constructed object as a Union command argument

### DISCUSSION THREAD

**Discussion by jelni** [Rapptz/discord.py#7168] [Q&A] [answered]
I want to use my custom converter that needs to be constructed with the Union type:
```py
@commands.command()
async def command(ctx, number: Union[Literal[False], ClampedNumber(1, 100)):
  ...
```
but this raises an error:
```
TypeError: Union[arg, ...]: each arg must be a type. Got <ClampedNumber object at 0x0000022EE87003A0>.
```
What's the best solution for this problem?

**Comment by Gobot1234** [comment, marked as answer]
I think the best solution here would be to add a method called `__call__` to your ClampedNumber class which should look like
```py
	def __call__(self):
		...
```
This will allow https://github.com/python/cpython/blob/bc3961485639cc73de7c4c7eed1b56f3c74939bf/Lib/typing.py#L173 to pass successfully, and allow your converter to be used as expected.

**Reply by jelni** [reply]
I also had to add the `__hash__` method and it works. Thank you!

**Comment by mikeshardmind** [comment]
Without knowledge of how your ClampedNumber class is implemented, it's a bit hard to make any recommendations here.

If the goal is just a number from 1-100 that plays nicely with typing's ecosystem

```py
class PositiveUnder101:

    @classmethod
    async def convert(cls, ctx, arg) -> int:
        # validate and return an int or raise here

```

Is simple and not subject to being broken by unknown future changes to python's typing ecosystem (some of which is currently up in the air with deferred evaluation of annotations being postponed)

There are more options available, but this

**Reply by jelni** [reply]
I don't really want to make another converter for every possible number combination, even if it doesn't work nicely with typing.
